Sources/XcodeGenKit/XCProjExtensions.swift

```diff
@@ -4,6 +4,59 @@ import XcodeProj
 
 extension PBXFileElement {
 
+    public struct NamePathPair: Comparable {
+        public var name: String?
+        public var path: String?
+
+        public static func < (lhs: PBXFileElement.NamePathPair, rhs: PBXFileElement.NamePathPair) -> Bool {
+            return lessThan(lhs, rhs, { $0 < $1 ? .orderedAscending : $0 == $1 ? .orderedSame : .orderedDescending })
+        }
+
+        public func localizedStandardLessThan(_ rhs: PBXFileElement.NamePathPair) -> Bool {
+            return NamePathPair.lessThan(self, rhs, { $0.localizedStandardCompare($1) })
+        }
+
+        private static func lessThan(
+            _ lhs: NamePathPair,
+            _ rhs: NamePathPair,
+            _ compare: (String, String) -> ComparisonResult
+        ) -> Bool {
+            return
+                lessThan(lhs.name ?? lhs.path, rhs.name ?? rhs.path, compare) ??
+                lessThan(lhs.name, rhs.name, compare) ??
+                lessThan(lhs.path, rhs.path, compare) ??
+                false
+        }
+
+        private static func lessThan(
+            _ lhs: String?,
+            _ rhs: String?,
+            _ compare: (String, String) -> ComparisonResult
+        ) -> Bool? {
+            switch (lhs, rhs) {
+            case let (lhs?, rhs?):
+                switch compare(lhs, rhs) {
+                case .orderedAscending:
+                    return true
+                case .orderedSame:
+                    return nil
+                case .orderedDescending:
+                    return false
+                }
+            case (nil, _?):
+                return false
+            case (_?, nil):
+                return true
+            case (nil, nil):
+                return nil
+            }
+        }
+    }
+
+    public var namePathPair: NamePathPair {
+        return NamePathPair(name: name, path: path)
```

Comment:
I think this could be simplified by having a simple `sortString` on `PBXFileElement`, which removes the need for the `NamePathPair`

```swift
var sortString: String { return "\(name ?? "")\(path ?? "")" }
```

Reply:
You could possibly have two pairs that have the same name/path pair, like `(name: "pa", path: "th"), and `(name: nil, path: "path")`. If you want to take this approach (which is a lot simpler), you'd have to introduce a character between the two that cannot be used in either name or path. I guess \t would suffice. I don't think you can use it in filenames (although I could be wrong). I'll modify the code to do that if you don't have any problems with that.

Reply:
Do you use tuples instead of the struct?

```swift
typealias Pair = (String, String)
func compare(_ lhs: Pair, _ rhs: Pair) -> ComparisonResult
```

tuples can't conform to any protocols, so you can use function to compare instead.

Reply:
I'm happy with using a simple search string. As you say @ChristopherRogers, you can add a random separator. Here's its `-`
```
var sortString: String { return "\(name ?? "")-\(path ?? "")" }
```

Reply:
You could create a duplicate with that, though. For example, ("a-b", "c") and ("a", "b-c") would both result in the sort string, "a-b-c".

Reply:
\t is fine then

Reply:
Or even
```
var sortString: String { return "name: \(name ?? ""), path: \(path ?? "")" }
```
It doesn’t really matter, it’s an implementation detail. Would be good to make it file private